### PR TITLE
Fixes .net graphing for project.assets.json target type

### PR DESCRIPTION
**Pull request by meghfossa**
# Overview

This PR introduces a change that ensures for `projectassetsjson` targets, 

- Project type dependencies are shrinked (and not reported)
- It correctly labels direct dependency based on `project.[framework]` data

## Acceptance criteria

- For target `projectassetsjson` type, dependencies of type project are not reported
- For target `projectassetsjson` type, dependencies are differentiated between direct and deep

## Testing plan

- Create .net project (or refer to any open source project.assets.json from github)
- Perform `fossa analyze -o | jq`, ensure produced dependency graphing,
   - reports all dependencies of package type
   - does not report dependencies of project type (if it has any transitive dependencies of package type, they are reported)
   - direct dependencies are correctly identified

## Risks

I did not like hack included with framework name matching, but I can't really move forward without having access to windows pc for complete testing. 

I do think, what I have should address 99+% cases. We still maintain previous behaviour (which was ignoring project's framework) when we are not able to find matching framework. 

## References

Closes https://github.com/fossas/team-analysis/issues/791
Closes https://github.com/fossas/team-analysis/issues/790

## Checklist

- [x] I added tests for this PR's change (or confirmed tests are not viable).
- [ ] If this PR introduced a user-visible change, I added documentation into `docs/`.
- [x] I updated `Changelog.md` if this change is externally facing. If this PR did not mark a release, I added my changes into an `# Unreleased` section at the top.
- [x] I updated `*schema.json` if I have made changes for `.fossa.yml`, `fossa-deps.{json, yaml, yml}`. You may also need to update these if you have added/removed new dependency (e.g. pip) or analysis target type (e.g. poetry).
- [x] I linked this PR to any referenced GitHub issues, if they exist.

### CODE REVIEW

File: src/Graphing.hs [hunk 1]
```diff
@@ -242,6 +243,31 @@ unfold seed getDeps toDependency = directs directNodes <> edges builtEdges
       let children = getDeps node
       map (node,) children ++ edgesFrom children
 
+-- | @unfoldDeep dep getDeps toDependency@ unfolds a graph, given:
+--
+-- - The @dep@ dependencies in the graph
```
Comment (cnr) [inline src/Graphing.hs:248]:
```suggestion
-- - The @deep@ dependencies in the graph
```

File: src/Strategy/NuGet/ProjectAssetsJson.hs [hunk 2]
```diff
@@ -68,52 +78,165 @@ analyze' file = do
...
+      map
+        (graphOfFramework $ projectFramework project)
+        (Map.toList $ targets project)
+
+graphOfFramework :: Map.Map Text (Set Text) -> (Text, Map.Map Text DependencyInfo) -> Graphing Dependency
+graphOfFramework projectFrameworkDeps (targetFramework, targetFrameworkDeps) =
+  withoutTags
+    . withoutProjects
+    . promoteToDirect isDirectDep
+    $ unfoldDeep allResolvedDeps getTransitiveDeps toDependency
```
Comment (cnr) [inline src/Strategy/NuGet/ProjectAssetsJson.hs:146]: The way I view the `Dependency` type is that once we've built something of type `Dependency`, we generally shouldn't touch it ever again

The fields on `Dependency` have very few guarantees, and depending on the tags field here to communicate type information indirectly feels brittle

When we need to modify `Dependency` values, it's usually a sign that we can model the domain more closely by using an intermediate type

We almost do that here with the `NuGetDep` type, but we eagerly convert it to `Dependency` with `toDependency`. May I recommend instead:

```hs
unfoldDeep allResolvedDeps getTransitiveDeps id :: Graphing NuGetDep
```

do the `Graphing.shrink withoutProjects` on that, slightly changing `withoutProjects`

```hs
withoutProjects :: Graphing NuGetDep -> Graphing NuGetDep
```

_then_ turn it into `Graphing Dependency` with `Graphing.gmap toDependency`

File: src/Strategy/NuGet/ProjectAssetsJson.hs [hunk 3]
```diff
@@ -68,52 +78,165 @@ analyze' file = do
...
+  deriving (Show, Eq, Ord)
 
-instance FromJSON ProjectAssetsJson where
-  parseJSON = withObject "ProjectAssetsJson" $ \obj ->
-    ProjectAssetsJson <$> obj .: "targets"
+data ProjectAssetsJson = ProjectAssetsJson
+  { targets :: Map.Map Text (Map.Map Text DependencyInfo)
+  , projectFramework :: Map.Map Text (Set Text)
+  }
+  deriving (Show, Eq, Ord)
```
Comment (cnr) [inline src/Strategy/NuGet/ProjectAssetsJson.hs:93]: [nit] there are a lot of `Text` values floating around; it's probably worthwhile to `newtype` wrapper some (or all?) of these. it was hard to follow some of the type signatures below and make sure the right text values were used

for json decoding purposes, you can derive `FromJSON` on the newtypes:

```hs
newtype FrameworkName = FrameworkName { unFrameworkName :: Text }
  deriving (Eq, Ord, Show, FromJSON)
```

File: test/NuGet/ProjectAssetsJsonSpec.hs [hunk 4]
```diff
@@ -54,16 +57,161 @@ dependencyFour =
...
+  Dependency
+    { dependencyType = NuGetType
+    , dependencyName = "five"
+    , dependencyVersion = Just (CEq "5.0.0")
+    , dependencyLocations = []
+    , dependencyEnvironments = mempty
+    , dependencyTags = Map.empty
+    }
+
+-- | Reference: https://github.com/NuGet/NuGet.Client/blob/820bfe9d5e5e4e7bf0b6310c285db7b07aff3087/test/NuGet.Core.Tests/NuGet.Frameworks.Test/NuGetFrameworkParseTests.cs
```
Comment (cnr) [inline test/NuGet/ProjectAssetsJsonSpec.hs:71]: [no action] wtf

File: test/NuGet/ProjectAssetsJsonSpec.hs [hunk 5]
```diff
@@ -54,16 +57,161 @@ dependencyFour =
...
+  , ("net452", ".NETFramework,Version=v4.5.2/win7-x86")
+  , ("net472", ".NETFramework,Version=4.7.2")
+  ]
+
 spec :: Spec
 spec = do
   testFile <- runIO (BS.readFile "test/NuGet/testdata/project.assets.json")
 
+  describe "approxEqlFramework" $ do
+    it "should return false when provided with equivalent framework" $ do
```
Comment (cnr) [inline test/NuGet/ProjectAssetsJsonSpec.hs:201]: is this test description correct?

**Review by cnr — APPROVED** [review]
LGTM generally; a few minor comments

File: test/NuGet/ProjectAssetsJsonSpec.hs (hunk 5 above)
Comment (meghfossa) [reply test/NuGet/ProjectAssetsJsonSpec.hs:201]: good catch!

File: src/Strategy/NuGet/ProjectAssetsJson.hs [hunk 6]
```diff
@@ -68,46 +82,94 @@ analyze' file = do
...
-    convertDep :: (Text, DependencyInfo) -> Maybe NuGetDep
-    convertDep (depString, dep) = case Text.splitOn "/" depString of
+    isProjectDep :: NuGetDep -> Bool
+    isProjectDep NuGetDep{..} = completeDepType == "project"
+
+    allResolvedDeps :: [NuGetDep]
+    allResolvedDeps = mapMaybe toNugetDep $ Map.toList targetFrameworkDeps
+
+    toNugetDep :: (Text, DependencyInfo) -> Maybe NuGetDep
+    toNugetDep (depString, dep) = case Text.splitOn "/" depString of
```
Comment (cnr) [inline src/Strategy/NuGet/ProjectAssetsJson.hs:165]: comment I forgot to add yesterday: we can do this check at JSON-decode-time, which would prevent us from silently ignoring invalid dependency keys -- and would also correctly propagate the error

there's a typeclass [called `FromJSONKey`](https://hackage.haskell.org/package/aeson-2.0.2.0/docs/Data-Aeson.html#t:FromJSONKey) that's used by `Map`'s `FromJSON` instance:

```hs
instance (FromJSONKey k, Ord k, FromJSON v) => FromJSON (Map k v) where
  ...
```

we can take advantage of this:

```hs
data NuGetDepKey = NuGetDepKey
  { depKeyName :: Text
  , depKeyVersion :: Text
  } deriving (Eq, Ord, Show)

instance FromJSONKey NuGetDepKey where
  fromJSONKey = FromJSONKeyTextParser $ \depString ->
    case Text.splitOn "/" depString of
      [name, ver] -> pure $ NuGetDepKey name ver
      other -> fail $ "Unexpected dependency key: " <> show depString
```

and change the type of `targets` in `ProjectAssetsJson`:

```hs
data ProjectAssetsJson = ProjectAssetsJson
  { targets :: Map.Map FrameworkName (Map.Map NuGetDepKey DependencyInfo)
  ...
```
